pkg/util/trivy/cache.go

```
@@ -186,3 +215,292 @@ func (c *TrivyCache) GetArtifact(id string) (types.ArtifactInfo, error) {
+func (c *PersistentCache) Keys() []string {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	keys := make([]string, c.lruCache.Len())
+	for i, key := range c.lruCache.Keys() {
+		keys[i] = key.(string)
+	}
+	return keys
+}
+
+// Len returns the number of items in the cache.
+func (c *PersistentCache) Len() int {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.lruCache.Len()
+}
+
+// Clear implements Cache#Clear.
+func (c *PersistentCache) Clear() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if err := c.db.Clear(); err != nil {
+		return err
+	}
+	c.lruCache.Purge()
+	c.currentCachedObjectTotalSize = 0
+	telemetry.SBOMCachedObjectSize.Set(0)
+	return nil
+}
+
+// removeOldest removes the least recently used item from the cache.
+func (c *PersistentCache) removeOldest() error {
+	key, ok := c.removeOldestKeyFromMemory()
+	if !ok {
+		return fmt.Errorf("in-memory cache is empty")
+	}
+
+	evicted := 0
+	if err := c.db.Delete([]string{key}, func(key string, value []byte) error {
+		evicted += len(value)
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	c.subCurrentCachedObjectTotalSize(evicted)
+
+	return nil
+}
+
+// RemoveOldest is a thread-safe version of removeOldest
+func (c *PersistentCache) RemoveOldest() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.removeOldest()
+}
+
+// reduceSize reduces the size of the cache to the target size by evicting the oldest items.
+func (c *PersistentCache) reduceSize(target int) error {
+	if c.currentCachedObjectTotalSize <= target {
+		return nil
+	}
+
+	prev := c.currentCachedObjectTotalSize
+	for c.currentCachedObjectTotalSize > target {
+		err := c.removeOldest()
+		if err != nil {
+			return err
+		}
+		if prev == c.currentCachedObjectTotalSize {
+			// if c.currentCachedObjectTotalSize is not updated by removeOldest then an item is stored in the lrucache without being stored in the local storage
+			return fmt.Errorf("cache and db are out of sync")
+		}
+	}
```

In order to check that every single call to `c.removeOldest()` leads to a modification of `c.currentCachedObjectTotalSize`, the line

```go
prev := c.currentCachedObjectTotalSize
```

needs to be moved inside the `for c.currentCachedObjectTotalSize > target {` loop.

---

Good catch ! I must have erased the line that updates prev by mistake. I wonder if we should just return an error here ? If the cache and db are out of sync, I think most operations will fail.